### ISSUE TITLE
Suggest missing Collate functionality

### ISSUE THREAD

**Issue by designermonkey** [ruflin/Elastica#1170]
Is there scope to add the collate functionality to Elastica? I can't seem to find how to do it so far.

**Comment by ruflin** [comment]
I assume this is the functionality you are referring to: https://www.elastic.co/guide/en/elasticsearch/reference/master/search-suggesters-phrase.html

There is no specific plan yet to add it but I would be really nice to have it in. PR's are always welcome.

**Comment by vgaldikas** [comment]
I would love to have a go at this. But I am new here. Any pointers where to begin? Maybe a link to pull request, that implements something similar to this, so I can analyze it?

**Comment by ruflin** [comment]
@vgaldikas There are already quite a few suggesters which can be found here: https://github.com/ruflin/Elastica/tree/master/lib/Elastica/Suggest Have a look at these and especially the Phrase one.

**Comment by vgaldikas** [comment]
The issue is not about suggester. I am aware of them, and I actually use it successfully. The issue is  about 'collate' functionality of Phrase suggester, I found a use case for it today at work, unfortunately Elastica does not provide API for it, and when I wrote the functionality as raw array, I got an exception saying it was not supported (I am not at work at the moment, so cannot provide the actual exception)

**Comment by ruflin** [comment]
I think what you need is to extend the Phrase suggester as you mentioned yourself: https://github.com/ruflin/Elastica/blob/master/lib/Elastica/Suggest/Phrase.php I haven't checked it in detail but it something like `setCollate($query AbstractQuery) { ... }` should get you started adding it to Elastica itself.

Curious to see what you tried and which errors you got.

**Comment by vgaldikas** [comment]
So at the moment my suggester works like so:
https://pastebin.com/WYbJEmAw

So the above works fine. Just one problem, is the suggestions contain phrases, which do not actually exist in the index. What I tried to do is:
https://pastebin.com/qhVxr83B

But as collate is not supported, I got the following exception:
Fatal error: Uncaught Elastica\Exception\ResponseException: suggester[phrase] doesn't support array field [collate] in /home/vytautas/dev/vytautas.kainos.dev/vendor/ruflin/elastica/lib/Elastica/Transport/Http.php on line 171
( ! ) Elastica\Exception\ResponseException: suggester[phrase] doesn't support array field [collate] in /home/vytautas/dev/vytautas.kainos.dev/vendor/ruflin/elastica/lib/Elastica/Transport/Http.php on line 171

**Comment by vgaldikas** [comment]
Oh crap. I just realised something, now after reading my last comment. Had I not used array, but constructed the query, via Query object, it probably would of worked, right?

**Comment by vgaldikas** [comment]
https://pastebin.com/RFmyXqs9

it seems to be working like so. 

Should there still be a way, to do it via something like, 'setCollate'?

**Comment by ruflin** [comment]
@vgaldikas I think it would definitively be nice to have something like `setCollate` to do this by passing a query to it. WDYT?

**Comment by vgaldikas** [comment]
@ruflin Yes I agree. Can I get a bit of guidance on how to develop. I forked and cloned your repository. I got the docker running. (https://pastebin.com/raw/zCrYzQcu). I can send request to elastic search, I can get into the shell via (make env-shell). 

But now I am lost... how do I actually start editing and running the code? I suppose I should set up a test, and create a testing index, and make it pass. But I am kind of lost... if I get into the env-shell, the file structure is diferent then what i have in my cloned repository:

https://pastebin.com/aiDga1SH

I suppose what I am looking for is what do I do once I have docker up and running in terms of where do I edit code, and how do I execute it. Sorry if this is a bit of noobish question, but my experience with docker is limited. Btw I am using Ubuntu, and PhpStorm.

**Comment by ruflin** [comment]
I edit the code directly on my machine and then run `make test TEST=...`. See https://github.com/ruflin/Elastica/blob/master/Makefile#L67 

You should first update the Phrase.php file and then add a test in the PhraseTest.php file. Best copy some code from existing functional tests.

**Comment by vgaldikas** [comment]
Whooray, I got it to work. 

Please see the comment on this commit: 	1139b8f. I am not sure how to best go about one of the parts, see if you could point me in right direction, because I can't seem to find a way to create an inline query properly.

**Comment by ruflin** [comment]
@vgaldikas Could you open a PR? This makes it easier to review and comment. 

I think you will need `AbstractQuery` instead of `Query` and then you can use `toArray()` on it. But easier to discuss on the PR when I see all changes together.
